lib/collection/src/shards/transfer/mod.rs

```diff
@@ -16,6 +27,64 @@ pub trait ShardTransferConsensus: Send + Sync {
     /// Returns `(commit, term)`.
     fn consensus_commit_term(&self) -> (u64, u64);
 
+    /// After snapshot recovery, propose to switch shard to `Partial`
+    ///
+    /// This is called after shard snapshot recovery has been completed on the remote. It submits a
+    /// proposal to consensus to switch the the shard state from `PartialSnapshot` to `Partial`.
+    ///
+    /// # Warning
+    ///
+    /// This only submits a proposal to consensus. Calling this does not guarantee that consensus
+    /// will actually apply the operation across the cluster.
+    fn snapshot_recovered_switch_to_partial(
+        &self,
+        transfer_config: &ShardTransfer,
+        collection_name: CollectionId,
+    ) -> CollectionResult<()>;
+
+    /// After snapshot recovery, propose to switch shard to `Partial` and confirm locally
+    ///
+    /// This is called after shard snapshot recovery has been completed on the remote. It submits a
+    /// proposal to consensus to switch the the shard state from `PartialSnapshot` to `Partial`.
+    ///
+    /// This method also confirms consensus applied the operation before returning by asserting the
+    /// change is propagated locally. If it fails, it will be retried for up to
+    /// `CONSENSUS_CONFIRM_RETRIES` times.
+    async fn snapshot_recovered_switch_to_partial_confirm(
+        &self,
+        transfer_config: &ShardTransfer,
+        collection_name: &str,
+        replica_set: &ShardReplicaSet,
+    ) -> CollectionResult<()> {
+        for attempts in (0..CONSENSUS_CONFIRM_RETRIES).rev() {
+            // Propose consensus operation
+            let proposal = self
+                .snapshot_recovered_switch_to_partial(transfer_config, collection_name.to_string());
+            match proposal {
+                Ok(()) => {}
+                Err(err) if attempts > 0 => {
+                    log::error!("Failed to propose snapshot recovered operation to consensus, retrying: {err}");
+                    continue;
+                }
+                Err(err) => return Err(err),
+            }
+
+            // Confirm local shard reached partial state
+            let confirm = replica_set
+                .wait_for_partial(transfer_config.to, CONSENSUS_CONFIRM_TIMEOUT)
+                .await;
+            match confirm {
+                Ok(()) => return Ok(()),
+                Err(err) if attempts > 0 => log::error!("Failed to confirm snapshot recovered operation on consensus, retrying: {err}"),
+                Err(err) => return Err(CollectionError::service_error(format!(
+                    "Failed to confirm snapshot recovered operation on consensus after {CONSENSUS_CONFIRM_RETRIES} retries: {err}"
+                ))),
+            }
+        }
```

Comment:
Called after the shard snapshot is recovered on the remote.

Propose to consensus to set the shard state from `PartialSnapshot` to `Partial`, but only if the transfer is still active. This is confirmed and retried up to 3 times before returning with `Ok(())`.